File: antia_numerical_methods/c/fun/linfitxy.c
/*	Least squares straight line fit when there is error in both x and y
        To fit equation of form y=a+b*x

        N : (input) Number of data points
        X : (input) Real array of length N containing the x values
        Y : (input) Real array of length N containing the y values
        SIGX : (input) estimated error in x values, assumed to be the
                same for all points
        SIGY : (input) estimated error in y values, assumed to be the
                same for all points
        RHO : (input) estimated correlation between errors in x and y
                assumed to be the same for all points
        XI : (output) Reall array of length N which will give the fitted x values
        YI : (output) Reall array of length N which will give the fitted y values
        A : (output) fitted value of the intercept
        B : (output) fitted value of the slope
        CHI : (output) value of chi^2 at the minimum

	Error status is returned by the value of the function LINFITXY.
        	0 value implies successful execution
                617 implies that discriminant of quadratic equation
                       is negative and calculations are aborted

	Required functions : none
*/

#include <math.h>

int linfitxy(int n, double *x, double *y, double sigx, double sigy,
	double rho, double *xi, double *yi, double *a, double *b,
	double *chi)
{
	int i;
	double r,s,sx,sy,sxy,sxx,syy,c0,c1,c2,del,b1,a1,t,t1,chi1;
	double x1,x2,y1,y2;

	r=rho*sigy/sigx; s=sigy*sigy/(sigx*sigx);
	sx=0.0; sy=0.0; sxy=0.0; sxx=0.0; syy=0.0;
	for(i=0; i<n; ++i){
		sx=sx+x[i]; sy=sy+y[i];
		sxy=sxy+x[i]*y[i]; sxx=sxx+x[i]*x[i]; syy=syy+y[i]*y[i];
	}
	sx=sx/n; sy=sy/n; sxy=sxy/n; sxx=sxx/n; syy=syy/n;

/*	Find the quadratic in the slope and solve it */
	c2=sigx*sigx*(sxy-sx*sy)+rho*sigx*sigy*(sx*sx-sxx);
	c1=sigx*sigx*(sy*sy-syy)+sigy*sigy*(sxx-sx*sx);
	c0=-rho*sigx*sigy*(sy*sy-syy)-sigy*sigy*(sxy-sx*sy);
	del=c1*c1-4*c0*c2;
	if(del<0.0) return 617;
	del=sqrt(del);
	if(c1>0.0) {b1=(-c1-del)/(2*c2);}
	else b1=(-c1+del)/(2*c2);
	*b=c0/(c2*b1);
	a1=sy-b1*sx;
	*a=sy-(*b)*sx;
	t1=(b1*r-s)/(b1-r);
	t=((*b)*r-s)/((*b)-r);

/*	Choose the solution with smaller chi^2 */
	chi1=0.0; *chi=0.0;
	for(i=0; i<n; ++i){
		x1=(t1*x[i]-y[i]+a1)/(t1-b1);
		x2=(t*x[i]-y[i]+(*a))/(t-(*b));
		y1=a1+b1*x1; y2=(*a)+(*b)*x2;
		chi1=chi1+(x[i]-x1)*(x[i]-x1)/(sigx*sigx)-2*rho*(x[i]-x1)*(y[i]-y1)
			/(sigx*sigy)+(y[i]-y1)*(y[i]-y1)/(sigy*sigy);
		*chi=*chi+(x[i]-x2)*(x[i]-x2)/(sigx*sigx)-2*rho*(x[i]-x2)*(y[i]-y2)
			/(sigx*sigy)+(y[i]-y2)*(y[i]-y2)/(sigy*sigy);
	}

	if(chi1< (*chi)){
		*chi=chi1; *a=a1; *b=b1; t=t1;
	}

	for(i=0; i<n; ++i){
		xi[i]=(t*x[i]-y[i]+(*a))/(t-(*b));
		yi[i]=(*a)+(*b)*xi[i];
	}
	return 0;
}

File: antia_numerical_methods@Beliavsky/f77/sub4/linfitxy.f
!       Least squares straight line fit when there is error in both x and y
!       To fit equation of form y=a+b*x
!
!       N : (input) Number of data points
!       X : (input) Real array of length N containing the x values
!       Y : (input) Real array of length N containing the y values
!       SIGX : (input) estimated error in x values, assumed to be the
!               same for all points
!       SIGY : (input) estimated error in y values, assumed to be the
!               same for all points
!       RHO : (input) estimated correlation between errors in x and y
!               assumed to be the same for all points
!       XI : (output) Reall array of length N which will give the fitted x values
!       YI : (output) Reall array of length N which will give the fitted y values
!       A : (output) fitted value of the intercept
!       B : (output) fitted value of the slope
!       CHI : (output) value of chi^2 at the minimum
!       IER : (output) Error parameter, IER=0 implies successful execution
!               IER=617 implies that discriminant of quadratic equation
!                       is negative and calculations are aborted
!
!       Required routines : none


      SUBROUTINE LINFITXY(N,X,Y,SIGX,SIGY,RHO,XI,YI,A,B,CHI,IER)
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(N),Y(N),XI(N),YI(N)

      R=RHO*SIGY/SIGX
      S=(SIGY/SIGX)**2
      SX=0.0
      SY=0.0
      SXY=0.0
      SXX=0.0
      SYY=0.0
      DO I=1,N
        SX=SX+X(I)
        SY=SY+Y(I)
        SXY=SXY+X(I)*Y(I)
        SXX=SXX+X(I)*X(I)
        SYY=SYY+Y(I)*Y(I)
      enddo
      SX=SX/N
      SY=SY/N
      SXY=SXY/N
      SXX=SXX/N
      SYY=SYY/N
      IER=0
      
!       Find the quadratic in the slope and solve it
      C2=SIGX*SIGX*(SXY-SX*SY)+RHO*SIGX*SIGY*(SX*SX-SXX)
      C1=SIGX*SIGX*(SY*SY-SYY)+SIGY*SIGY*(SXX-SX*SX)
      C0=-RHO*SIGX*SIGY*(SY*SY-SYY)-SIGY*SIGY*(SXY-SX*SY)
      DEL=C1*C1-4*C0*C2
      IF(DEL.LT.0.0) THEN
        IER=617
        RETURN
      ENDIF
      DEL=SQRT(DEL)
      IF(C1.GT.0.0) THEN
        B1=(-C1-DEL)/(2*C2)
      ELSE
        B1=(-C1+DEL)/(2*C2)
      ENDIF
      B=C0/(C2*B1)
      A1=SY-B1*SX
      A=SY-B*SX
      T1=(B1*R-S)/(B1-R)
      T=(B*R-S)/(B-R)

!       Choose the solution with smaller chi^2
      CHI1=0.0
      CHI=0.0
      DO I=1,N
        X1=(T1*X(I)-Y(I)+A1)/(T1-B1)
        X2=(T*X(I)-Y(I)+A)/(T-B)
        Y1=A1+B1*X1
        Y2=A+B*X2
        CHI1=CHI1+((X(I)-X1)/SIGX)**2-2*RHO*(X(I)-X1)*(Y(I)-Y1)
     1          /(SIGX*SIGY) + ((Y(I)-Y1)/SIGY)**2
        CHI=CHI+((X(I)-X2)/SIGX)**2-2*RHO*(X(I)-X2)*(Y(I)-Y2)
     1          /(SIGX*SIGY) + ((Y(I)-Y2)/SIGY)**2
      ENDDO
      
      IF(CHI1.LT.CHI) THEN
        CHI=CHI1
        A=A1
        B=B1
        T=T1
      ENDIF

      DO I=1,N
        XI(I)=(T*X(I)-Y(I)+A)/(T-B)
        YI(I)=A+B*XI(I)
      ENDDO
      END
